src/librustc_mir/hair/cx/expr.rs

```diff
@@ -652,13 +652,8 @@ fn make_mirror_unadjusted<'a, 'gcx, 'tcx>(cx: &mut Cx<'a, 'gcx, 'tcx>,
                 user_ty,
             );
 
-            // Check to see if this cast is a "coercion cast", where the cast is actually done
-            // using a coercion (or is a no-op).
-            let cast = if let Some(&TyCastKind::CoercionCast) =
-                cx.tables()
-                .cast_kinds()
-                .get(source.hir_id)
-            {
+            let source_ty = cx.tables().expr_ty(source);
+            let cast = if source_ty == expr_ty {
```

Comment:
I misunderstood what the old code did. The old code also checked for casts e.g. from `&T` to `*const T` or from `&[u8; 42]` to `&[u8]`. I think for now we can change the `cast_kinds` field to a `coercion_casts` field which is a `HashSet<HirId>` that is set for every coercion cast. (instead of storing the entire `TyCastKind::CoercionCast` in there)

Reply:
@oli-obk Tried doing some refactoring for this:

> I think for now we can change the `cast_kinds` field to a `coercion_casts` field which is a `HashSet<HirId>` that is set for every coercion cast

But there are a few places where it looks like we need `cast_kinds`

https://github.com/rust-lang/rust/blob/ea43c3c688980edd6f09a4cb632c9eb996c4f2af/src/librustc_typeck/check/writeback.rs#L358-L372

https://github.com/rust-lang/rust/blob/ea43c3c688980edd6f09a4cb632c9eb996c4f2af/src/librustc_typeck/check/cast.rs#L428-L439

How do we proceed?